src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs

```diff
@@ -3,24 +3,41 @@
 using Microsoft.Extensions.Logging;
 using System.Collections;
 
-namespace CarbonAware.Aggregators.CarbonAware
+namespace CarbonAware.Aggregators.CarbonAware;
+
+public class CarbonAwareAggregator : ICarbonAwareAggregator
 {
-    public class CarbonAwareAggregator : ICarbonAwareAggregator
+    private readonly ILogger<CarbonAwareAggregator> _logger;
+
+    private readonly ICarbonAware _plugin;
+
+    public CarbonAwareAggregator(ILogger<CarbonAwareAggregator> logger, ICarbonAware plugin)
     {
-        private readonly ILogger<CarbonAwareAggregator> _logger;
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _plugin = plugin ?? throw new ArgumentNullException(nameof(plugin));
+    }
 
-        private readonly ICarbonAware _plugin;
+    public async Task<double> CalcEmissionsAverageAsync(IDictionary props)
+    {
+        ValidateAverageProps(props);
+        var list = await GetEmissionsDataAsync(props);
+        var value = list.Any() ? list.Select(x => x.Rating).Average() : 0;
+        _logger.LogInformation($"Carbon Intensity Average: {value}");
+        return value;
+    }
 
-        public CarbonAwareAggregator(ILogger<CarbonAwareAggregator> logger, ICarbonAware plugin)
+    private void ValidateAverageProps(IDictionary props)
+    {
+        if (!props.Contains(CarbonAwareConstants.Locations) ||
+            !props.Contains(CarbonAwareConstants.Start) ||
+            !props.Contains(CarbonAwareConstants.End))
         {
-            this._logger = logger ?? throw new ArgumentNullException(nameof(logger));
-            this._plugin = plugin;
+            throw new ArgumentException("Missing properties to calculate average");
```

Comment:
🐑 in the future we might want to split this out so that we can explicitly tell users which props they're missing

Reply:
Agree. I'll wait till #26 lands to make other changes that affects this.